Bad span for ? removal suggestion in parentheses

### Code

```Rust
fn foo() -> Option<()> {
    let x = Some(());
    (x?)
}
```


### Current output

```Shell
warning: unnecessary parentheses around block return value
 --> src/lib.rs:3:5
  |
3 |     (x?)
  |     ^  ^
  |
  = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
  |
3 -     (x?)
3 +     x?
  |

error[E0308]: `?` operator has incompatible types
 --> src/lib.rs:3:6
  |
3 |     (x?)
  |      ^^ expected `Option<()>`, found `()`
  |
  = note: `?` operator cannot convert from `()` to `Option<()>`
  = note:   expected enum `Option<()>`
          found unit type `()`
help: try removing this `?`
  |
3 -     (x?)
3 +     (x?
  |
help: try wrapping the expression in `Some`
  |
3 |     (Some(x?))
  |      +++++  +

For more information about this error, try `rustc --explain E0308`.
```


### Desired output

```Shell
...
help: try removing this `?`
  |
3 -     (x?)
3 +     (x)
```


### Rationale and extra context

The suggestion span for removing the `?` is off-by-one, and suggests removing the parenthesis instead of the `?`.


### Other cases

_No response_

### Anything else?

rustc 1.73.0-nightly (7bd81ee19 2023-07-13)
binary: rustc
commit-hash: 7bd81ee1902c049691d0a1f03be5558bee51d100
commit-date: 2023-07-13
host: aarch64-apple-darwin
release: 1.73.0-nightly
LLVM version: 16.0.5

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -764,7 +764,7 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                             Some(ty) if expected == ty => {
                                 let source_map = self.tcx.sess.source_map();
                                 err.span_suggestion(
-                                    source_map.end_point(cause.span),
+                                    source_map.end_point(cause.span()),
                                     "try removing this `?`",
                                     "",
                                     Applicability::MachineApplicable,

--- a/tests/ui/suggestions/remove-question-symbol-with-paren.rs
+++ b/tests/ui/suggestions/remove-question-symbol-with-paren.rs
@@ -1,0 +1,9 @@
+// https://github.com/rust-lang/rust/issues/114392
+
+fn foo() -> Option<()> {
+    let x = Some(());
+    (x?)
+    //~^ ERROR `?` operator has incompatible types
+}
+
+fn main() {}

--- a/tests/ui/suggestions/remove-question-symbol-with-paren.stderr
+++ b/tests/ui/suggestions/remove-question-symbol-with-paren.stderr
@@ -1,0 +1,22 @@
+error[E0308]: `?` operator has incompatible types
+  --> $DIR/remove-question-symbol-with-paren.rs:5:6
+   |
+LL |     (x?)
+   |      ^^ expected `Option<()>`, found `()`
+   |
+   = note: `?` operator cannot convert from `()` to `Option<()>`
+   = note:   expected enum `Option<()>`
+           found unit type `()`
+help: try removing this `?`
+   |
+LL -     (x?)
+LL +     (x)
+   |
+help: try wrapping the expression in `Some`
+   |
+LL |     (Some(x?))
+   |      +++++  +
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

